master/internal/api_master.go

```diff
@@ -25,3 +29,65 @@ func (a *apiServer) GetMaster(
 		Config:    configStruct,
 	}, err
 }
+
+// effectiveOffset returns effective offset.
+func effectiveOffset(reqOffset int, totalItems int) (offset int) {
+	switch {
+	case reqOffset < -totalItems:
+		return 0
+	case reqOffset < 0:
+		return totalItems + reqOffset
+	default:
+		return reqOffset
+	}
+}
+
+// effectiveLimit returns effective limit.
+// Input: Limit 0 is treated as no limit
+// Output: Limit -1 is treated as no limit
+func effectiveLimit(reqLimit int, offset int, totalItems int) (limit int) {
+	limit = -1
+	if reqLimit != 0 {
+		limit = reqLimit
+		if limit > totalItems-offset {
+			limit = totalItems - offset
+		}
+	}
+	return limit
+}
```

Comment:
non-blocking: could probably drop the comment and write this similarly to above.
```suggestion
func effectiveLimit(limit int, offset int, total int) int {
	switch {
	case limit == 0:
		return -1
	case limit > total - offset:
		return total - offset
	default:
		return limit
	}
}
```